packages/alfa-html/src/alphabet.ts

```diff
@@ -237,6 +264,522 @@ const tagName: Pattern = (stream, emit, state) => {
+      stream.advance(1);
+      state.temporaryBuffer = "";
+      return scriptDataEndTagOpen;
+    case Char.ExclamationMark:
+      stream.advance(1);
+      emit({ type: TokenType.Character, data: "<" });
+      emit({ type: TokenType.Character, data: "!" });
+      return scriptDataEscapeStart;
+    default:
+      emit({ type: TokenType.Character, data: "<" });
+      return scriptData;
+  }
+};
+
+/**
+ * 8.2.4.16
+ * @see https://www.w3.org/TR/html/syntax.html#tokenizer-script-data-end-tag-open-state
+ */
+const scriptDataEndTagOpen: Pattern = (stream, emit, state) => {
+  const char = stream.peek(0);
+  if (char === null || !isAlpha(char)) {
+    emit({ type: TokenType.Character, data: "<" });
+    emit({ type: TokenType.Character, data: "/" });
+    return scriptData;
+  }
+
+  state.tag = {
+    type: TokenType.EndTag,
+    name: ""
+  };
+  return scriptDataEndTagName;
+};
+
+/**
+ * 8.2.4.17
+ * @see https://www.w3.org/TR/html/syntax.html#tokenizer-script-data-end-tag-name-state
+ */
+const scriptDataEndTagName: Pattern = (stream, emit, state) => {
+  function isAppropriateEndTagToken(): boolean {
+    return state.tag !== null && state.tag.name === "script";
+  }
+
+  function anythingElse() {
+    emit({ type: TokenType.Character, data: "<" });
+    emit({ type: TokenType.Character, data: "/" });
+    for (let i = 0; i < state.temporaryBuffer.length; i++) {
+      emit({ type: TokenType.Character, data: state.temporaryBuffer[i] });
+    }
+    return scriptData;
+  }
+
+  const char = stream.peek(0);
+  if (char === null) {
+    return anythingElse;
+  }
+
+  switch (char) {
+    case Char.CharacterTabulation:
+    case Char.LineFeed:
+    case Char.FormFeed:
+    case Char.Space:
+      stream.advance(1);
+      if (isAppropriateEndTagToken()) {
+        return beforeAttributeName;
+      }
+      return anythingElse;
+    case Char.Solidus:
+      stream.advance(1);
+      if (isAppropriateEndTagToken()) {
+        return selfClosingStartTag;
+      }
+      return anythingElse;
+    case Char.GreaterThanSign:
+      stream.advance(1);
+      if (isAppropriateEndTagToken()) {
+        emit(state.tag!);
+        return data;
+      }
+      return anythingElse;
+    default:
+      if (isUpperCase(char)) {
+        stream.advance(1);
+        if (state.tag !== null) {
+          state.tag.name += fromCharCode(char).toLowerCase();
+        } else {
+          state.tag = {
+            type: TokenType.EndTag,
+            name: fromCharCode(char).toLowerCase()
+          };
+        }
+        state.temporaryBuffer += fromCharCode(char);
+        break;
+      }
+      if (isLowerCase(char)) {
+        stream.advance(1);
+        if (state.tag !== null) {
+          state.tag.name += fromCharCode(char);
+        } else {
+          state.tag = {
+            type: TokenType.EndTag,
+            name: fromCharCode(char)
+          };
+        }
+        state.temporaryBuffer += fromCharCode(char);
+        break;
+      }
```

Comment:
The above can be simplified quite a bit:

```ts
if (isAlpha(char)) {
  stream.advance(1);

  let str = fromCharCode(char);

  state.tag!.name += str.toLowerCase();
  state.temporaryBuffer += str;
  break;
}
```

`state.tag` will always be different from `null` when this pattern is executed, which is why I make use of a non-null assertion (`!`). The casing of the character also doesn't really matter; if it's alphabetic then we simply lowercase it.

Reply:
Duly noted. I've changed the let statement to const, as the value is never reassigned.